src/Unsafe/MutableArray.hs

```diff
@@ -23,42 +31,57 @@ import GHC.Exts
 -- # Unsafe wrappers
 ----------------------------
 
+-- | A mutable array holding @a@s
 type MutArr# a = MutableArray# RealWorld a
 
+-- | Get the size of a mutable array
 sizeMutArr :: MutArr# a -> Int
 sizeMutArr arr  = I# (sizeofMutableArray# arr)
 
+-- | Given a size, try to allocate a mutable array
+-- of this size. The size should be greater than zero.
 newMutArr :: Int -> a -> MutArr# a
 newMutArr (I# size) x =
   case newArray of
     (# _, array #) -> array
   where
     newArray = runRW# $ \stateRW -> newArray# size x stateRW
 
+-- | Write to a given mutable array arr, given an
+-- index in [0, size(arr)-1] , and a value
 writeMutArr :: MutArr# a -> Int -> a -> ()
 writeMutArr mutArr (I# ix) val =
   case doWrite of _ -> ()
   where
     doWrite = runRW# $ \stateRW -> writeArray# mutArr ix val stateRW
 
+-- | Read from a given mutable array arr, given an index
+-- in [0, size(arr)-1]
 readMutArr :: MutArr# a -> Int -> a
 readMutArr mutArr (I# ix) =
   case doRead of (# _, a #) -> a
   where
     doRead = runRW# $ \stateRW -> readArray# mutArr ix stateRW
 
--- | Resize a mutable array, using the first value to fill in the new cells
+-- | Grow a mutable array. This is given an array, a given larger size,
+-- and it returns a larger array of the given size using the first value
+-- to fill in the new cells and copying over all the unchanged cells. This
+-- fails for a size smaller than the size of the given array.
```

Comment:
It's ok. And so are the other functions in this module, I'm just picking one randomly. But I'm a big fan of writing documentation like this:

```
-- | @'resizeMutArr' arr size@, where @size@ is larger than the length of @arr@, returns a new array with length @size@ […]
```
That is: give name to things, and refer to them precisely, rather than using semi-precise English phrases to work around the fact that you didn't give yourself names. So, in the future, I want to see more documentation like this.

It's particularly easy/reasonable for data structures like this, since everything tends to have a super clear semantics, so you can give a pretty formal description of the function as its docstring.